src/protocol/context.rs

```diff
@@ -1,46 +1,85 @@
 #[derive(Debug)]
-pub struct ProtocolContext<'a, S: SpaceIndex, F> {
-    pub participant: &'a Participant<S>,
-    pub gateway: &'a Gateway<S, F>,
-    pub identity: Identity,
+pub struct ProtocolContext<'a, N> {
+    role: Identity,
+    step: UniqueStepId,
+    participant: &'a Participant,
+    gateway: &'a Gateway<N>,
 }
 
-impl<'a, S: Step + SpaceIndex, F: Network<S>> ProtocolContext<'a, S, F> {
-    pub fn new(
-        participant: &'a Participant<S>,
-        gateway: &'a Gateway<S, F>,
-        identity: Identity,
-    ) -> Self {
+impl<'a, N> ProtocolContext<'a, N> {
+    pub fn new(role: Identity, participant: &'a Participant, gateway: &'a Gateway<N>) -> Self {
         Self {
+            role,
+            step: UniqueStepId::default(),
             participant,
             gateway,
-            identity,
         }
     }
 
+    /// The role of this context.
+    #[must_use]
+    pub fn role(&self) -> Identity {
+        self.role
+    }
+
+    /// A unique identifier for this stage of the protocol execution.
+    #[must_use]
+    pub fn step(&self) -> &UniqueStepId {
+        &self.step
+    }
+
+    /// Make a sub-context.
+    /// Note that each invocation of this should use a unique value of `step`.
+    #[must_use]
+    pub fn narrow<S: Step>(&self, step: &S) -> Self {
+        ProtocolContext {
+            role: self.role,
+            step: self.step.narrow(step),
+            participant: self.participant,
+            gateway: self.gateway,
+        }
+    }
+
+    /// Get the PRSS instance for this step.
+    #[must_use]
+    pub fn prss(&self) -> &PrssSpace {
+        self.participant.prss(&self.step)
+    }
+}
+
+impl<N: Network> ProtocolContext<'_, N> {
+    /// Get a set of communications channels to different peers.
+    #[must_use]
+    pub fn mesh(&self) -> Mesh<'_, '_, N> {
+        self.gateway.mesh(&self.step)
+    }
+
     /// Request multiplication for a given record. This function is intentionally made async
     /// to allow backpressure if infrastructure layer cannot keep up with protocols demand.
     /// In this case, function returns only when multiplication for this record can actually
     /// be processed.
     #[allow(clippy::unused_async)] // eventually there will be await b/c of backpressure implementation
-    pub async fn multiply(&'a self, record_id: RecordId, step: S) -> SecureMul<'a, S, F> {
-        SecureMul::new(&self.participant[step], self.gateway, step, record_id)
+    pub async fn multiply(&self, record_id: RecordId) -> SecureMul<'_, N> {
+        SecureMul::new(self.prss(), self.gateway, &self.step, record_id)
     }
 
     /// Request reveal for a given record.
     #[allow(clippy::unused_async)] // eventually there will be await b/c of backpressure implementation
-    pub fn reveal(&'a self, record_id: RecordId, step: S) -> Reveal<'a, F, S> {
-        Reveal::new(self.gateway, step, record_id)
+    #[must_use]
+    pub fn reveal(&self, record_id: RecordId) -> Reveal<'_, '_, N> {
```

Comment:
Should this similarly be made async for the same back-pressure mechanism as stated above?

Not really related to this diff in particular, but just something that occurs to me.